File: web/js/core/board.js
import { BOARD_SIZE, Team, PieceType } from "./constants.js";
import { Pawn, Rook, Knight, Bishop, Queen, King } from "./piece.js";

export class Board {
  constructor() {
    this._grid = Array.from({ length: BOARD_SIZE }, () => Array(BOARD_SIZE).fill(null));
    this._setup();
  }

  _setup() {
    const back = [Rook, Knight, Bishop, Queen, King, Bishop, Knight, Rook];
    for (let c = 0; c < 8; c++) {
      this._grid[0][c] = new back[c](Team.BLACK);
      this._grid[7][c] = new back[c](Team.WHITE);
      this._grid[1][c] = new Pawn(Team.BLACK);
      this._grid[6][c] = new Pawn(Team.WHITE);
    }
  }

  get(row, col)           { return this._grid[row][col]; }
  set(row, col, piece)    { this._grid[row][col] = piece; }

  applyMove([sr, sc], [dr, dc]) {
    const captured = this._grid[dr][dc];
    this._grid[dr][dc] = this._grid[sr][sc];
    this._grid[sr][sc] = null;
    if (this._grid[dr][dc]) this._grid[dr][dc].hasMoved = true;
    return captured;
  }

  undoMove([sr, sc], [dr, dc], captured, movedBefore) {
    this._grid[sr][sc] = this._grid[dr][dc];
    this._grid[dr][dc] = captured;
    if (this._grid[sr][sc]) this._grid[sr][sc].hasMoved = movedBefore;
  }

  findKing(team) {
    for (let r = 0; r < 8; r++)
      for (let c = 0; c < 8; c++) {
        const p = this._grid[r][c];
        if (p?.team === team && p.type === PieceType.KING) return [r, c];
      }
    return null;
  }

  allPieces(team) {
    const result = [];
    for (let r = 0; r < 8; r++)
      for (let c = 0; c < 8; c++) {
        const p = this._grid[r][c];
        if (p?.team === team) result.push({ pos: [r, c], piece: p });
      }
    return result;
  }

  isInCheck(team) {
    const kingPos = this.findKing(team);
    if (!kingPos) return false;
    const opp = team === Team.WHITE ? Team.BLACK : Team.WHITE;
    for (const { pos, piece } of this.allPieces(opp)) {
      if (piece.legalMoves(pos, this).some(([r,c]) => r===kingPos[0] && c===kingPos[1]))
        return true;
    }
    return false;
  }
}

File: src/core/board.py
"""8×8 board — holds pieces and handles move execution."""
from __future__ import annotations
from typing import Optional, List, Tuple
from src.core.constants import (
    BOARD_SIZE, WHITE, BLACK,
    PAWN, ROOK, KNIGHT, BISHOP, QUEEN, KING,
)
from src.core.position import Position
from src.core.piece import Piece, Pawn, Rook, Knight, Bishop, Queen, King


class Board:
    def __init__(self):
        self._grid: List[List[Optional[Piece]]] = [
            [None] * BOARD_SIZE for _ in range(BOARD_SIZE)
        ]
        self._setup()

    # ------------------------------------------------------------------
    # Setup
    # ------------------------------------------------------------------

    def _setup(self):
        back_row = [Rook, Knight, Bishop, Queen, King, Bishop, Knight, Rook]

        for col, cls in enumerate(back_row):
            self._grid[0][col] = cls(BLACK)
            self._grid[7][col] = cls(WHITE)

        for col in range(BOARD_SIZE):
            self._grid[1][col] = Pawn(BLACK)
            self._grid[6][col] = Pawn(WHITE)

    # ------------------------------------------------------------------
    # Access
    # ------------------------------------------------------------------

    def get(self, pos: Position) -> Optional[Piece]:
        return self._grid[pos.row][pos.col]

    def set(self, pos: Position, piece: Optional[Piece]):
        self._grid[pos.row][pos.col] = piece

    def find_king(self, team: str) -> Optional[Position]:
        for row in range(BOARD_SIZE):
            for col in range(BOARD_SIZE):
                p = self._grid[row][col]
                if p and p.piece_type == KING and p.team == team:
                    return Position(row, col)
        return None

    def all_pieces(self, team: str) -> List[Tuple[Position, Piece]]:
        result = []
        for row in range(BOARD_SIZE):
            for col in range(BOARD_SIZE):
                p = self._grid[row][col]
                if p and p.team == team:
                    result.append((Position(row, col), p))
        return result

    # ------------------------------------------------------------------
    # Move execution
    # ------------------------------------------------------------------

    def apply_move(self, src: Position, dst: Position) -> Optional[Piece]:
        """Move piece from src→dst, return any captured piece."""
        piece = self.get(src)
        captured = self.get(dst)
        self.set(dst, piece)
        self.set(src, None)
        if piece:
            piece.has_moved = True
        return captured

    def is_in_check(self, team: str) -> bool:
        king_pos = self.find_king(team)
        if king_pos is None:
            return False
        opponent = BLACK if team == WHITE else WHITE
        for pos, piece in self.all_pieces(opponent):
            if king_pos in piece.legal_moves(pos, self):
                return True
        return False

    # ------------------------------------------------------------------
    # Display
    # ------------------------------------------------------------------

    def render(self) -> str:
        lines = ["  a b c d e f g h"]
        for row in range(BOARD_SIZE):
            rank = 8 - row
            cells = []
            for col in range(BOARD_SIZE):
                p = self._grid[row][col]
                cells.append(p.symbol() if p else ".")
            lines.append(f"{rank} {' '.join(cells)}")
        return "\n".join(lines)
